File: tests/fixtures/scala/type_inference/control_flow/control_flow.scala
sealed trait Shape
case class Circle(radius: Double) extends Shape
case class Rectangle(width: Double, height: Double) extends Shape
case object Unknown extends Shape

object ControlFlowDemo {
  def describe(x: Any): String = {
    if (x.isInstanceOf[String]) {
      x.asInstanceOf[String].toUpperCase
    } else if (x.isInstanceOf[Int]) {
      s"number: $x"
    } else {
      "unknown"
    }
  }

  def matchShape(shape: Shape): String = shape match {
    case c: Circle => s"circle r=${c.radius}"
    case r: Rectangle => s"rect ${r.width}x${r.height}"
    case _ => "unknown"
  }

  def matchValue(x: Any): String = x match {
    case s: String => s.toUpperCase
    case n: Int => n.toString
    case _ => "other"
  }

  def main(args: Array[String]): Unit = {
    println(describe("hello"))
    println(describe(42))
    println(matchShape(Circle(1.0)))
    println(matchValue("test"))
  }
}

File: tests/fixtures/dart/type_inference/control_flow/control_flow.dart
sealed class Result {
  const Result();
}

class Success extends Result {
  String value = '';
}

class Failure extends Result {
  String message = '';
}

String handleIs(Object value) {
  if (value is String) {
    return value.toUpperCase();
  }
  if (value is int) {
    return 'number: $value';
  }
  return 'unknown';
}

String handleResult(Result result) {
  if (result is Success) {
    return result.value;
  }
  if (result is Failure) {
    return result.message;
  }
  return 'pending';
}

String handleKind(String kind, Object payload) {
  if (kind == 'text') {
    return payload.toString();
  }
  return 'other';
}

void main() {
  print(handleIs('hello'));
  print(handleIs(42));
  print(handleResult(Success()));
  print(handleKind('text', 'payload'));
}
